In [ ]:
import pandas as pd

url = 'https://eds-217-essential-python.github.io/data/national_parks.csv'

df = pd.read_csv(url)

Have fun reading this amazingly well put together analysis on busy national parks. Pleas enjoy 

 EM



 BTW..... The Red Clay Stays - Moment of Truth = 9.9/10

In [18]:
df.head()

,year,gnis_id,geometry,metadata,number_of_records,parkname,region,state,unit_code,unit_name,unit_type,visitors
0,1904,1163670,POLYGON,NaN,1,Crater Lake,PW,OR,CRLA,Crater Lake National Park,National Park,1500.0
1,1941,1531834,MULTIPOLYGON,NaN,1,Lake Roosevelt,PW,WA,LARO,Lake Roosevelt National Recreation Area,National Recreation Area,0.0
2,1961,2055170,MULTIPOLYGON,NaN,1,Lewis and Clark,PW,WA,LEWI,Lewis and Clark National Historical Park,National Historical Park,69000.0
3,1935,1530459,MULTIPOLYGON,NaN,1,Olympic,PW,WA,OLYM,Olympic National Park,National Park,2200.0
4,1982,277263,POLYGON,NaN,1,Santa Monica Mountains,PW,CA,SAMO,Santa Monica Mountains National Recreation Area,National Recreation Area,468144.0


### Important to know:

df[df['column'] > value] # to filter the pattern

df.sort_values('column', ascending = False).head(n) # the top N pattern

In [19]:
df[df['unit_type'] == "National Park"]

,year,gnis_id,geometry,metadata,number_of_records,parkname,region,state,unit_code,unit_name,unit_type,visitors
0,1904,1163670,POLYGON,NaN,1,Crater Lake,PW,OR,CRLA,Crater Lake National Park,National Park,1500.0
3,1935,1530459,MULTIPOLYGON,NaN,1,Olympic,PW,WA,OLYM,Olympic National Park,National Park,2200.0
5,1919,578853,MULTIPOLYGON,NaN,1,NaN,NE,ME,ACAD,Acadia National Park,National Park,64000.0
8,1944,1377082,POLYGON,NaN,1,NaN,IM,TX,BIBE,Big Bend National Park,National Park,1409.0
22,1948,293666,POLYGON,NaN,1,NaN,SE,FL,EVER,Everglades National Park,National Park,7482.0
...,...,...,...,...,...,...,...,...,...,...,...,...
21555,1905,1163670,POLYGON,NaN,1,Crater Lake,PW,OR,CRLA,Crater Lake National Park,National Park,1400.0
21556,1905,1609331,POLYGON,http://nrdata.nps.gov/programs/Lands/YELL_meta...,1,Yellowstone,IM,WY,YELL,Yellowstone National Park,National Park,26188.0
21557,1905,255948,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/KICA_meta...,1,Kings Canyon,PW,CA,KICA,Kings Canyon National Park,National Park,1000.0
21558,1905,1528416,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/MORA_meta...,1,Mount Rainier,PW,WA,MORA,Mount Rainier National Park,National Park,928.0


How many rows describe units whose unit_type is exactly National Park?

    - there are 4,743 rows

In [20]:
df[df['year'] == "Total"]

,year,gnis_id,geometry,metadata,number_of_records,parkname,region,state,unit_code,unit_name,unit_type,visitors
300,Total,2767391,POLYGON,http://nrdata.nps.gov/programs/Lands/HONO_meta...,1,Honouliuli,PW,HI,HONO,Honouliuli National Monument,National Monument,NaN
308,Total,2766708,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/MIIN_meta...,1,Minidoka,PW,WA,MIIN,Minidoka National Historic Site,National Historic Site,NaN
337,Total,2035517,POLYGON,http://nrdata.nps.gov/programs/Lands/VALL_meta...,1,Valles Caldera,IM,NM,VALL,Valles Caldera National Preserve,Valles Caldera National Preserve,NaN
369,Total,2766905,POLYGON,http://nrdata.nps.gov/programs/Lands/PULL_meta...,1,Pullman,MW,IL,PULL,Pullman National Monument,National Monument,NaN
386,Total,1163670,POLYGON,NaN,1,Crater Lake,PW,OR,CRLA,Crater Lake National Park,National Park,34669486.0
...,...,...,...,...,...,...,...,...,...,...,...,...
763,Total,1556735,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/HAFE_meta...,1,Harpers Ferry,NC,WV,HAFE,Harpers Ferry National Historical Park,National Historical Park,32575872.0
764,Total,634820,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/PIRO_meta...,1,Pictured Rocks,MW,MI,PIRO,Pictured Rocks National Lakeshore,National Lakeshore,19983228.0
765,Total,1888945,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/MOJA_META...,1,Mojave,PW,CA,MOJA,Mojave National Preserve,National Preserve,10594537.0
766,Total,74272,POLYGON,http://nrdata.nps.gov/programs/Lands/BUFF_meta...,1,Buffalo,MW,AR,BUFF,Buffalo National River,National River,38574595.0


How many rows have a year of Total?

    - 386 rows

In [21]:
by_year = df[df['year'] != "Total"]
by_year.shape


(21174, 12)

Use ~ or != to build a table called by_year that has none of them in it. How many rows does by_year have

    - 21174

In [22]:
list_of_codes = ['PW', 'IM', 'AK']

by_year[by_year['region'].isin(list_of_codes)]

,year,gnis_id,geometry,metadata,number_of_records,parkname,region,state,unit_code,unit_name,unit_type,visitors
0,1904,1163670,POLYGON,NaN,1,Crater Lake,PW,OR,CRLA,Crater Lake National Park,National Park,1500.0
1,1941,1531834,MULTIPOLYGON,NaN,1,Lake Roosevelt,PW,WA,LARO,Lake Roosevelt National Recreation Area,National Recreation Area,0.0
2,1961,2055170,MULTIPOLYGON,NaN,1,Lewis and Clark,PW,WA,LEWI,Lewis and Clark National Historical Park,National Historical Park,69000.0
3,1935,1530459,MULTIPOLYGON,NaN,1,Olympic,PW,WA,OLYM,Olympic National Park,National Park,2200.0
4,1982,277263,POLYGON,NaN,1,Santa Monica Mountains,PW,CA,SAMO,Santa Monica Mountains National Recreation Area,National Recreation Area,468144.0
...,...,...,...,...,...,...,...,...,...,...,...,...
21553,1906,1528416,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/MORA_meta...,1,Mount Rainier,PW,WA,MORA,Mount Rainier National Park,National Park,1786.0
21555,1905,1163670,POLYGON,NaN,1,Crater Lake,PW,OR,CRLA,Crater Lake National Park,National Park,1400.0
21556,1905,1609331,POLYGON,http://nrdata.nps.gov/programs/Lands/YELL_meta...,1,Yellowstone,IM,WY,YELL,Yellowstone National Park,National Park,26188.0
21557,1905,255948,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/KICA_meta...,1,Kings Canyon,PW,CA,KICA,Kings Canyon National Park,National Park,1000.0


The region column uses two-letter codes. Build a list of the codes PW, IM, and AK, and use .isin() to filter by_year down to units in those three regions. How many rows?

    - There are 9,894 rows.

In [23]:
by_year[(by_year['unit_type'] == "National Park") & (by_year['visitors'] > 5000000)]

,year,gnis_id,geometry,metadata,number_of_records,parkname,region,state,unit_code,unit_name,unit_type,visitors
794,2016,1018241,MULTIPOLYGON,NaN,1,NaN,SE,NC,GRSM,Great Smoky Mountains National Park,National Park,11312786.0
1033,2016,23866,POLYGON,http://nrdata.nps.gov/programs/Lands/GRCA_META...,1,Grand Canyon,IM,AZ,GRCA,Grand Canyon National Park,National Park,5969811.0
1087,2016,255923,POLYGON,http://nrdata.nps.gov/programs/Lands/YOSE_META...,1,Yosemite,PW,CA,YOSE,Yosemite National Park,National Park,5028868.0
1172,2015,1018241,MULTIPOLYGON,NaN,1,NaN,SE,NC,GRSM,Great Smoky Mountains National Park,National Park,10712674.0
1410,2015,23866,POLYGON,http://nrdata.nps.gov/programs/Lands/GRCA_META...,1,Grand Canyon,IM,AZ,GRCA,Grand Canyon National Park,National Park,5520736.0
1548,2014,1018241,MULTIPOLYGON,NaN,1,NaN,SE,NC,GRSM,Great Smoky Mountains National Park,National Park,10099276.0
1924,2013,1018241,MULTIPOLYGON,NaN,1,NaN,SE,NC,GRSM,Great Smoky Mountains National Park,National Park,9354695.0
2297,2012,1018241,MULTIPOLYGON,NaN,1,NaN,SE,NC,GRSM,Great Smoky Mountains National Park,National Park,9685829.0
2670,2011,1018241,MULTIPOLYGON,NaN,1,NaN,SE,NC,GRSM,Great Smoky Mountains National Park,National Park,9008830.0
3039,2010,1018241,MULTIPOLYGON,NaN,1,NaN,SE,NC,GRSM,Great Smoky Mountains National Park,National Park,9463538.0


Using two conditions in one filter on by_year, find the rows that are National Parks and recorded more than 5 million visitors. How many are there, and what does each row represent? Use by_year rather than parks, because the Total rows you removed in question 2 would otherwise be counted alongside the single years.

    - there are 59 rows.

## Part 2. RANKING

In [24]:
df.sort_values('visitors', ascending=False).head(10)

,year,gnis_id,geometry,metadata,number_of_records,parkname,region,state,unit_code,unit_name,unit_type,visitors
397,Total,1009494,MULTIPOLYGON,NaN,1,NaN,NT,NC,BLRI,Blue Ridge Parkway,Parkway,871922828.0
674,Total,255952,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/GOGA_meta...,1,Golden Gate,PW,CA,GOGA,Golden Gate National Recreation Area,National Recreation Area,611031225.0
412,Total,1018241,MULTIPOLYGON,NaN,1,NaN,SE,NC,GRSM,Great Smoky Mountains National Park,National Park,521947058.0
724,Total,123648,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/NATR_meta...,1,Natchez Trace,SE,MS,NATR,Natchez Trace Parkway,Parkway,443145232.0
653,Total,858265,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/LAKE_META...,1,Lake Mead,PW,NV,LAKE,Lake Mead National Recreation Area,National Recreation Area,411700377.0
507,Total,584569,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/GWMP_meta...,1,George Washington Memorial Parkway,NC,VA,GWMP,George Washington Memorial Parkway,National Parkway,330201962.0
690,Total,882894,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/GATE_meta...,1,Gateway,NE,NY,GATE,Gateway National Recreation Area,National Recreation Area,329664174.0
727,Total,1478412,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/COLO_meta...,1,Colonial,NE,VA,COLO,Colonial National Historical Park,National Historical Park,282420671.0
599,Total,531835,POLYGON,http://nrdata.nps.gov/programs/Lands/LINC_meta...,1,Lincoln Memorial,NC,DC,LINC,Lincoln Memorial,National Memorial,238620382.0
676,Total,2738287,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/NACA_meta...,1,National Capital Parks,NC,DC,NACA,National Capital Parks,Park,236187845.0


Rank all of parks by visitors, largest first, and look at the top 10. Something is wrong with this answer. Say what, and say why in one sentence.

    - The problem is that this is the total amount of vistors for the whole year.

In [25]:
by_year.sort_values('visitors', ascending=False).head()

,year,gnis_id,geometry,metadata,number_of_records,parkname,region,state,unit_code,unit_name,unit_type,visitors
11164,1987,255952,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/GOGA_meta...,1,Golden Gate,PW,CA,GOGA,Golden Gate National Recreation Area,National Recreation Area,21767176.0
10853,1988,255952,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/GOGA_meta...,1,Golden Gate,PW,CA,GOGA,Golden Gate National Recreation Area,National Recreation Area,21759271.0
11475,1986,255952,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/GOGA_meta...,1,Golden Gate,PW,CA,GOGA,Golden Gate National Recreation Area,National Recreation Area,21582368.0
5923,2002,1009494,MULTIPOLYGON,NaN,1,NaN,NT,NC,BLRI,Blue Ridge Parkway,Parkway,21538760.0
12942,1981,255952,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/GOGA_meta...,1,Golden Gate,PW,CA,GOGA,Golden Gate National Recreation Area,National Recreation Area,20279789.0


Fix it, and rank by_year instead. Which single park unit had the highest visitor count in any one year, and in what year?

    - Golden Gate had the highest visitor count in the year 1987.

In [26]:
by_year[by_year['year'] == "2016"].sort_values(by = 'visitors', ascending=False)[['unit_name', 'state', 'visitors']].head(10)

,unit_name,state,visitors
1055,Golden Gate National Recreation Area,CA,15638777.0
779,Blue Ridge Parkway,NC,15175578.0
794,Great Smoky Mountains National Park,NC,11312786.0
889,George Washington Memorial Parkway,VA,10323339.0
1070,Gateway National Recreation Area,NY,8651770.0
980,Lincoln Memorial,DC,7915934.0
1034,Lake Mead National Recreation Area,NV,7175891.0
1033,Grand Canyon National Park,AZ,5969811.0
1103,Natchez Trace Parkway,MS,5891315.0
977,Vietnam Veterans Memorial,DC,5299713.0


Filter to National Parks in the year 2016, then find the top 10 by visitors. Show only the unit_name, state, and visitors columns.

    - Look up :)


In [27]:
np_2016 = by_year[(by_year['year'] =='2016') & (by_year["unit_type"] == "National Park")]
np_2016.sort_values(by = 'visitors', ascending=False)[['unit_name', 'state', 'visitors']]

print(np_2016.shape)



(61, 12)


In [28]:
np_2016[np_2016['visitors'] > 1000000]

,year,gnis_id,geometry,metadata,number_of_records,parkname,region,state,unit_code,unit_name,unit_type,visitors
771,2016,1530459,MULTIPOLYGON,NaN,1,Olympic,PW,WA,OLYM,Olympic National Park,National Park,3390221.0
773,2016,578853,MULTIPOLYGON,NaN,1,NaN,NE,ME,ACAD,Acadia National Park,National Park,3303393.0
794,2016,1018241,MULTIPOLYGON,NaN,1,NaN,SE,NC,GRSM,Great Smoky Mountains National Park,National Park,11312786.0
808,2016,1488805,MULTIPOLYGON,NaN,1,NaN,NE,VA,SHEN,Shenandoah National Park,National Park,1437341.0
848,2016,1455716,POLYGON,http://nrdata.nps.gov/programs/Lands/ARCH_meta...,1,Arches,IM,UT,ARCH,Arches National Park,National Park,1585718.0
873,2016,1450197,POLYGON,http://nrdata.nps.gov/programs/Lands/BRCA_meta...,1,Bryce Canyon,IM,UT,BRCA,Bryce Canyon National Park,National Park,2365110.0
878,2016,1609331,POLYGON,http://nrdata.nps.gov/programs/Lands/YELL_meta...,1,Yellowstone,IM,WY,YELL,Yellowstone National Park,National Park,4257177.0
881,2016,365210,POLYGON,http://nrdata.nps.gov/programs/Lands/HAVO_meta...,1,Hawai'i Volcanoes,PW,HI,HAVO,Hawai'i Volcanoes National Park,National Park,1887580.0
892,2016,266000,POLYGON,http://nrdata.nps.gov/programs/Lands/SEQU_meta...,1,Sequoia,PW,CA,SEQU,Sequoia National Park,National Park,1254688.0
900,2016,1597571,POLYGON,http://nrdata.nps.gov/programs/Lands/GRTE_meta...,1,Grand Teton,IM,WY,GRTE,Grand Teton National Park,National Park,3270076.0


Your np_2016 table has one row per National Park that reported a 2016 figure. Print its shape to see how many parks that is, then find how many of them recorded more than a million visitors.

    - There are 61 rows total but only 22 of them have > 1,000,000 visitors

In [29]:
most_visits = np_2016['visitors'].idxmax()

print('most visited park', np_2016.loc[most_visits, 'unit_name'])

most visited park Great Smoky Mountains National Park


Use .idxmax() and .loc[] on np_2016 to print the name of the most-visited National Park of 2016 on its own, without displaying the whole row.

    - The most visited park of 2016 was Great Smoky Mountains National Park

### Part 3. Put them together

In [30]:
state_list = ['CA', 'UT', 'AZ']


state_park_new = np_2016[np_2016['state'].isin(state_list)].sort_values('visitors', ascending = False)
state_park_new




,year,gnis_id,geometry,metadata,number_of_records,parkname,region,state,unit_code,unit_name,unit_type,visitors
1033,2016,23866,POLYGON,http://nrdata.nps.gov/programs/Lands/GRCA_META...,1,Grand Canyon,IM,AZ,GRCA,Grand Canyon National Park,National Park,5969811.0
1087,2016,255923,POLYGON,http://nrdata.nps.gov/programs/Lands/YOSE_META...,1,Yosemite,PW,CA,YOSE,Yosemite National Park,National Park,5028868.0
981,2016,1455157,POLYGON,http://nrdata.nps.gov/programs/Lands/ZION_meta...,1,Zion,IM,UT,ZION,Zion National Park,National Park,4295127.0
1009,2016,277261,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/jotr_meta...,1,Joshua Tree,PW,CA,JOTR,Joshua Tree National Park,National Park,2505286.0
873,2016,1450197,POLYGON,http://nrdata.nps.gov/programs/Lands/BRCA_meta...,1,Bryce Canyon,IM,UT,BRCA,Bryce Canyon National Park,National Park,2365110.0
848,2016,1455716,POLYGON,http://nrdata.nps.gov/programs/Lands/ARCH_meta...,1,Arches,IM,UT,ARCH,Arches National Park,National Park,1585718.0
1028,2016,857501,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/DEVA_META...,1,Death Valley,PW,CA,DEVA,Death Valley National Park,National Park,1296283.0
892,2016,266000,POLYGON,http://nrdata.nps.gov/programs/Lands/SEQU_meta...,1,Sequoia,PW,CA,SEQU,Sequoia National Park,National Park,1254688.0
1003,2016,1434866,POLYGON,http://nrdata.nps.gov/programs/Lands/CARE_meta...,1,Capitol Reef,IM,UT,CARE,Capitol Reef National Park,National Park,1064904.0
1100,2016,10683,MULTIPOLYGON,http://nrdata.nps.gov/programs/Lands/SAGU_meta...,1,Saguaro,IM,AZ,SAGU,Saguaro National Park,National Park,820426.0


Build a list of the state codes CA, UT, and AZ. Filter np_2016 to National Parks in those three states, rank them by visitors, and display the whole ranked list.

    - look up :)



In [31]:
state_park_new['state'].value_counts()


state
CA    9
UT    5
AZ    3
Name: count, dtype: int64

Which of those three states has the park at the top of your list, and which has the most parks in it? Answer both in a markdown cell. You can count the parks per state by eye from your ranked table.

    - The park with the most visitors is Grand Canyon, located in AZ. CA has the most parks in the list, with 9 parks total.

In [32]:
np_2016.sort_values('visitors', ascending=False).head(1)

,year,gnis_id,geometry,metadata,number_of_records,parkname,region,state,unit_code,unit_name,unit_type,visitors
794,2016,1018241,MULTIPOLYGON,NaN,1,NaN,SE,NC,GRSM,Great Smoky Mountains National Park,National Park,11312786.0


 answer this: a colleague asks you which national park is “the busiest”. What do you need to know before you can answer, and what would you tell them? Use at least two numbers from today’s work

    - Before I could say which park is "the busiest," I'd need to know what year/time period and what metric we mean by busy. A metric such as the total annual visitors, or average daily visitors are very important to consider. Based on today's ranking, Great Smoky Mountains National Park had the highest raw visitor count at 11,300,000 visitors in 2016, but that doesn't account for park size or seasonality, so it may not be the most crowded. There are many ground truths that need to be considered in order to make an accurate determination. 



    KIT...